modules/cudacodec/src/video_decoder.cpp

```diff
@@ -57,23 +57,42 @@ void cv::cudacodec::detail::VideoDecoder::create(const FormatInfo& videoFormat)
                                             cudaVideoCreate_PreferCUVID;
 
     // Validate video format.  These are the currently supported formats via NVCUVID
-    CV_Assert(cudaVideoCodec_MPEG1 == _codec ||
-              cudaVideoCodec_MPEG2 == _codec ||
-              cudaVideoCodec_MPEG4 == _codec ||
-              cudaVideoCodec_VC1   == _codec ||
-              cudaVideoCodec_H264  == _codec ||
-              cudaVideoCodec_JPEG  == _codec ||
-              cudaVideoCodec_YUV420== _codec ||
-              cudaVideoCodec_YV12  == _codec ||
-              cudaVideoCodec_NV12  == _codec ||
-              cudaVideoCodec_YUYV  == _codec ||
-              cudaVideoCodec_UYVY  == _codec );
+	CV_Assert(  cudaVideoCodec_MPEG1    == _codec ||
+                cudaVideoCodec_MPEG2    == _codec ||
+                cudaVideoCodec_MPEG4    == _codec ||
+                cudaVideoCodec_VC1      == _codec ||
+                cudaVideoCodec_H264     == _codec ||
+                cudaVideoCodec_JPEG     == _codec ||
+                cudaVideoCodec_H264_SVC == _codec ||
+                cudaVideoCodec_H264_MVC == _codec ||
+                cudaVideoCodec_HEVC     == _codec ||
+                cudaVideoCodec_VP8      == _codec ||
+                cudaVideoCodec_VP9      == _codec ||
+                cudaVideoCodec_YUV420   == _codec ||
+                cudaVideoCodec_YV12     == _codec ||
+                cudaVideoCodec_NV12     == _codec ||
+                cudaVideoCodec_YUYV     == _codec ||
+                cudaVideoCodec_UYVY     == _codec);
 
     CV_Assert(cudaVideoChromaFormat_Monochrome == _chromaFormat ||
               cudaVideoChromaFormat_420        == _chromaFormat ||
               cudaVideoChromaFormat_422        == _chromaFormat ||
               cudaVideoChromaFormat_444        == _chromaFormat);
 
+    // Check video format is supported by GPU's hardware video decoder
+    CUVIDDECODECAPS decodeCaps = {};
+    decodeCaps.eCodecType = _codec;
+    decodeCaps.eChromaFormat = _chromaFormat;
+    decodeCaps.nBitDepthMinus8 = videoFormat.nBitDepthMinus8;
+    cuSafeCall(cuvidGetDecoderCaps(&decodeCaps));
```

Comment:
API version check is required to avoid issues like [this](https://devtalk.nvidia.com/default/topic/1020133/nvapi/-cuvidgetdecodercaps-can-t-found-in-libnvcuvid-so-cuda-version-is-375-20/).

Reply:
Looking at the CUDA headers it appears that prior to CUDA 9.0, the available formats are dictated by the CUDA distribution in either cuviddec.h or dynlink_cuviddec.h.
From CUDA 9.0 onward the Nvidia Codec SDK needs to be used.  From the documentation it looks as though the only version of the Nvidia Codec SDK with libs and headers is version 9.0 which supports all the available codecs and the cuvidGetDecoderCaps() function.
Is it therefore acceptable to use the CUDART_VERSION to dictate the level of support in the following manner

>         // Validate video format.  These are the currently supported formats via NVCUVID
>         bool codecSupported =   cudaVideoCodec_MPEG1    == _codec ||
>                                 cudaVideoCodec_MPEG2    == _codec ||
>                                 cudaVideoCodec_MPEG4    == _codec ||
>                                 cudaVideoCodec_VC1      == _codec ||
>                                 cudaVideoCodec_H264     == _codec ||
>                                 cudaVideoCodec_JPEG     == _codec ||
>                                 cudaVideoCodec_H264_SVC == _codec ||
>                                 cudaVideoCodec_H264_MVC == _codec ||
>                                 cudaVideoCodec_YV12     == _codec ||
>                                 cudaVideoCodec_NV12     == _codec ||
>                                 cudaVideoCodec_YUYV     == _codec ||
>                                 cudaVideoCodec_UYVY     == _codec;
> 
>     #if defined (HAVE_CUDA)
>     #if (CUDART_VERSION >= 6500)  
>         codecSupported |=       cudaVideoCodec_HEVC     == _codec;
>     #endif
>     #if  ((CUDART_VERSION == 7500) || (CUDART_VERSION >= 9000)) 
>         codecSupported |=       cudaVideoCodec_VP8      == _codec ||
>                                 cudaVideoCodec_VP9      == _codec ||
>                                 cudaVideoCodec_YUV420   == _codec;
>     #endif
>     #endif
> 
>         CV_Assert(codecSupported);
>         CV_Assert(  cudaVideoChromaFormat_Monochrome == _chromaFormat ||
>                     cudaVideoChromaFormat_420        == _chromaFormat ||
>                     cudaVideoChromaFormat_422        == _chromaFormat ||
>                     cudaVideoChromaFormat_444        == _chromaFormat);
> 
>     #if defined (HAVE_CUDA) && (CUDART_VERSION >= 9000)
>         // Check video format is supported by GPU's hardware video decoder
>         CUVIDDECODECAPS decodeCaps = {};
>         decodeCaps.eCodecType = _codec;
>         decodeCaps.eChromaFormat = _chromaFormat;
>         decodeCaps.nBitDepthMinus8 = videoFormat.nBitDepthMinus8;
>         cuSafeCall(cuvidGetDecoderCaps(&decodeCaps));
>         if (!decodeCaps.bIsSupported)
>             CV_Error(Error::StsUnsupportedFormat, "Video source is not supported by hardware video decoder");
> 
>         CV_Assert(  videoFormat.width   >= decodeCaps.nMinWidth &&
>                     videoFormat.height  >= decodeCaps.nMinHeight &&
>                     videoFormat.width   <= decodeCaps.nMaxWidth &&
>                     videoFormat.height  <= decodeCaps.nMaxHeight);
>     #endif

Reply:
Looks good! Thank you for investigation.

Do we really need `#if defined (HAVE_CUDA)` check? (cuda modules should have this definition enabled).

Reply:
Yes your right, I thought cudacodec could be used independently of CUDA but of course it can't, due to the dependence on GpuMat for the decoded frame.

I have also re-enabled the performance test which was disabled with defined(HAVE_VIDEO_INPUT) which I can't find reference to anywhere is this redundant or should I have left it in and made the change somewhere else?